trackintel/preprocessing/staypoints.py

```diff
@@ -72,27 +84,34 @@ def generate_locations(
             db = DBSCAN(eps=epsilon, min_samples=num_samples, algorithm="ball_tree", metric=distance_metric)
 
         if agg_level == "user":
-            location_id_counter = 0
-            # TODO: change into groupby
-            for user_id_this in ret_stps["user_id"].unique():
-                # Slice staypoints array by user. This is not a copy!
-                user_staypoints = ret_stps[ret_stps["user_id"] == user_id_this]
 
-                if distance_metric == "haversine":
-                    # the input is converted to list of (lat, lon) tuples in radians unit
-                    p = np.array([[radians(g.y), radians(g.x)] for g in user_staypoints.geometry])
-                else:
-                    p = np.array([[g.x, g.y] for g in user_staypoints.geometry])
-                labels = db.fit_predict(p)
+            # we create a new column: 'serial_num' where we assign serial numbers to each user id
+            ret_stps = ret_stps.assign(serial_num=(ret_stps["user_id"]).astype("category").cat.codes)
+            if print_progress:
+                tqdm.pandas(desc="User location generation")
+                ret_stps = ret_stps.groupby("user_id", as_index=False).progress_apply(
+                    _generate_locations_per_user,
+                    geo_col=geo_col,
+                    distance_metric=distance_metric,
+                    db=db,
+                    num_total_stps=len(ret_stps),
+                )
+            else:
+                ret_stps = ret_stps.groupby("user_id", as_index=False).apply(
+                    _generate_locations_per_user,
+                    geo_col=geo_col,
+                    distance_metric=distance_metric,
+                    db=db,
+                    num_total_stps=len(ret_stps),
+                )
+
+            ret_stps = ret_stps.drop(columns={"serial_num"})
```

Comment:
For MOBIS dataset, the returned location id is `[-2146247740 -2146247739 -2146247738 ...  2147416528  2147416529  2147416530]`, this means an integer overflow. We can of course change the offset into int64, but I think this is not elegant and we will again face this issue when dealing with real big data (MOBIS "only" contains 4k users).

I could not think of a way to `reduce` offset while still ensure uniqueness. Otherwise, I can try to imitate the way we did for the other functions and add to this pull request.

Reply:
Ah nice, good that you have found it now. Overflow errors are often very hard to detect later on. 
I had a look and maybe I have found a solution to this problem. 
The solution has two steps:
1. Delete [line 207](https://github.com/mie-lab/trackintel/pull/232/files#diff-e5f965436547ae93edf1ce55e8085c31f842c8a131e70edf674692b546205b9eR207) in `_generate_locations_per_user`
2. Add the following code after [line 108](https://github.com/mie-lab/trackintel/pull/232/files#diff-e5f965436547ae93edf1ce55e8085c31f842c8a131e70edf674692b546205b9eR108)
```python 
# sort so that the last location id of a user = max(location id)
ret_stps = ret_stps.sort_values(['user_id', 'location_id'])
# identify start positions of new user_ids
start_of_user_id = ret_stps['user_id'] != ret_stps['user_id'].shift(1)
# calculate the offset (= last location id of the previous user)
# multiplication is to mask all positions where no new user starts and addition is to have a +1 when a 
# new user starts
loc_id_offset = ret_stps['location_id'].shift(1) * start_of_user_id + start_of_user_id
# fill first nan with 0 and create the cumulative sum
loc_id_offset = loc_id_offset.fillna(0).cumsum()
ret_stps['location_id2'] = ret_stps['location_id'] + loc_id_offset
```
@abcnishant007 let me know if you can work with this.

Reply:
Thanks @henrymartin1, this works perfect. I have applied these changes, needed to apply a sort by "user_id" if handling shuffled rows. Apart from that, this works as is. Tested for 
- geolife with 100K lines (first 100K lines extracted from the geolife combined data sent by @hong2223 )
- "with_strings_for_user_id.csv" file as discussed previously in this chain